scripts/pages/zoning/zoning.ts

```diff
@@ -0,0 +1,922 @@
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const n = this.panels.pointsList.Children().indexOf(point);
+		const index = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		this.selectedZone.zone?.regions[index].points.splice(n, 1);
+		point.DeleteAsync(0);
+	}
+
+	static addRegionPoint(point) {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const index = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		if (index > -1) {
+			const i = this.selectedZone.zone.regions[index].points.length;
+			this.selectedZone.zone.regions[index].points.push([point.x, point.y]);
+			this.addPointToList(i, [point.x, point.y]);
+		}
+	}
+
+	static setRegionBottom() {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		const bottom = Number.parseFloat(this.panels.regionBottom.text);
+		region.bottom = Number.isNaN(bottom) ? 0 : bottom;
+	}
+
+	static setRegionHeight() {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		const height = Number.parseFloat(this.panels.regionHeight.text);
+		region.height = Number.isNaN(height) ? 0 : height;
+	}
+
+	static setRegionSafeHeight() {
+		if (!this.selectedZone || !this.selectedZone.zone) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		const height = Number.parseFloat(this.panels.regionSafeHeight.text);
+		region.regionSafeHeight = Number.isNaN(height) ? 0 : height;
+	}
+
+	static updateRegionTPDest() {
+		if (!this.selectedZone || !this.selectedZone.zone || !this.teleDestList) return;
+
+		const regionIndex = this.panels.regionSelect.GetSelected().GetAttributeInt('value', -1);
+		const teleDestIndex = this.panels.regionTPDest.GetSelected()?.GetAttributeInt('value', 0);
+		const region = this.selectedZone.zone.regions[regionIndex];
+		if (teleDestIndex === 0) {
+			// no teleport destination for this region
+			region.teleDestTargetname = '';
+			//@ts-expect-error property must be optional
+			delete region.teleDestPos;
+			//@ts-expect-error property must be optional
+			delete region.teleDestYaw;
+
+			this.panels.regionTPPos.x.text = '';
+			this.panels.regionTPPos.y.text = '';
+			this.panels.regionTPPos.z.text = '';
+			this.panels.regionTPYaw.text = '';
+			this.setRegionTPDestTextEntriesActive(false);
+		} else if (teleDestIndex === 1 && region.points.length > 0) {
+			if (!region.teleDestPos || !region.teleDestYaw) {
+				region.teleDestPos = [] as number[];
+				region.teleDestPos = [0, 0, region.bottom];
+				for (const [_, point] of region.points.entries()) {
```

Comment:
Just an array, just iterate over region.points
If you wanna be nerdy use `region.points[0].reduce((total, curr) => total += curr) / region.points.length)`, up to you

Reply:
we're only talking about a few points at a time (usually 4), but wouldn't this require iterating over the array twice? Once for each x and once for each y?

Reply:
True, fair